# <span style='color:magenta; font-weight:bold;'>Exploratory Data Analysis (EDA)</span>

In this section, we explore the processed data to:
1. Understand the structure and distribution of the data.
2. Identify missing or inconsistent values.
3. Detect duplicates or outliers.
4. Gain insights to guide the data cleaning and transformation process.

### Data Sources
The datasets used in this analysis were ingested from APIs and external sources, then transformed using Python scripts. The processed data is stored in the `data/processed/` directory and loaded here for exploration.

### Steps:
- Display the first few rows of each dataset.
- Check for missing values and duplicates.
- Summarize key statistics for numerical and categorical columns.


### Import Libraries & Load the Data

In [1]:
# Essentials
import pandas as pd
import numpy as np
import os

In [2]:
# Load datasets
products = pd.read_csv("../data/processed/product_dimension.csv")
categories = pd.read_csv("../data/processed/category_dimension.csv")
carts = pd.read_csv("../data/processed/cart_dimension.csv", parse_dates=["cart_date"])
users = pd.read_csv("../data/processed/user_dimension.csv")
sales = pd.read_csv("../data/processed/sales_fact_table.csv")

### General EDA function

In [3]:
def eda(df):
    print("-------------------------------TOP 5 RECORDS-----------------------------")
    display(df.head())
    
    print("\n-------------------------------INFO--------------------------------------")
    display(df.info())
    
    print("\n-------------------------------Describe----------------------------------")
    display(df.describe())
    
    print("\n-------------------------------Columns-----------------------------------")
    display(df.columns)
    
    print("\n----------------------------Missing Values-------------------------------")
    display(df.isnull().sum())
    
    print("\n--------------------------Shape Of Data---------------------------------")
    display(df.shape)

### EDA on Datasets

In [4]:
# Exploring the carts dataset
print("=================================Carts Data=================================")
eda(carts)

# Exploring the products dataset
print("=================================Products Data=================================")
eda(products)

# Exploring the categories dataset
print("=================================Categories=================================")
eda(categories)

# Exploring the users dataset
print("=================================Users Data=================================")
eda(users)

# Exploring the Sales Fact Table dataset
print("=================================Sales Data=================================")
eda(sales)

=================================Carts Data=================================
-------------------------------TOP 5 RECORDS-----------------------------


cart_id                 cart_date
0        1 2020-03-02 00:00:00+00:00
1        2 2020-01-02 00:00:00+00:00
2        3 2020-03-01 00:00:00+00:00
3        4 2020-01-01 00:00:00+00:00
4        5 2020-03-01 00:00:00+00:00


-------------------------------INFO--------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype              
---  ------     --------------  -----              
 0   cart_id    7 non-null      int64              
 1   cart_date  7 non-null      datetime64[ns, UTC]
dtypes: datetime64[ns, UTC](1), int64(1)
memory usage: 244.0 bytes


None


-------------------------------Describe----------------------------------


cart_id
count  7.000000
mean   4.000000
std    2.160247
min    1.000000
25%    2.500000
50%    4.000000
75%    5.500000
max    7.000000


-------------------------------Columns-----------------------------------


Index(['cart_id', 'cart_date'], dtype='object')


----------------------------Missing Values-------------------------------


cart_id      0
cart_date    0
dtype: int64


--------------------------Shape Of Data---------------------------------


(7, 2)

=================================Products Data=================================
-------------------------------TOP 5 RECORDS-----------------------------


product_id   price                                       product_name  \
0           1  109.95  Fjallraven - Foldsack No. 1 Backpack, Fits 15 ...   
1           2   22.30             Mens Casual Premium Slim Fit T-Shirts    
2           3   55.99                                 Mens Cotton Jacket   
3           4   15.99                               Mens Casual Slim Fit   
4           5  695.00  John Hardy Women's Legends Naga Gold & Silver ...   

                                         description  \
0  Your perfect pack for everyday use and walks i...   
1  Slim-fitting style, contrast raglan long sleev...   
2  great outerwear jackets for Spring/Autumn/Wint...   
3  The color could be slightly different between ...   
4  From our Legends Collection, the Naga was insp...   

                                           image_url  
0  https://fakestoreapi.com/img/81fPKd-2AYL._AC_S...  
1  https://fakestoreapi.com/img/71-3HjGNDUL._AC_S...  
2  https://fakestoreapi.com/img/71li-ujtlUL._AC_U...  
3  https://fakestoreapi.com/img/71YXzeOuslL._AC_U...  
4  https://fakestoreapi.com/img/71pWzhdJNwL._AC_U...


-------------------------------INFO--------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   product_id    20 non-null     int64  
 1   price         20 non-null     float64
 2   product_name  20 non-null     object 
 3   description   20 non-null     object 
 4   image_url     20 non-null     object 
dtypes: float64(1), int64(1), object(3)
memory usage: 932.0+ bytes


None


-------------------------------Describe----------------------------------


product_id       price
count    20.00000   20.000000
mean     10.50000  162.046000
std       5.91608  272.220532
min       1.00000    7.950000
25%       5.75000   15.240000
50%      10.50000   56.490000
75%      15.25000  110.962500
max      20.00000  999.990000


-------------------------------Columns-----------------------------------


Index(['product_id', 'price', 'product_name', 'description', 'image_url'], dtype='object')


----------------------------Missing Values-------------------------------


product_id      0
price           0
product_name    0
description     0
image_url       0
dtype: int64


--------------------------Shape Of Data---------------------------------


(20, 5)

=================================Categories=================================
-------------------------------TOP 5 RECORDS-----------------------------


category_id     category_name
0            1       electronics
1            2          jewelery
2            3    men's clothing
3            4  women's clothing


-------------------------------INFO--------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   category_id    4 non-null      int64 
 1   category_name  4 non-null      object
dtypes: int64(1), object(1)
memory usage: 196.0+ bytes


None


-------------------------------Describe----------------------------------


category_id
count     4.000000
mean      2.500000
std       1.290994
min       1.000000
25%       1.750000
50%       2.500000
75%       3.250000
max       4.000000


-------------------------------Columns-----------------------------------


Index(['category_id', 'category_name'], dtype='object')


----------------------------Missing Values-------------------------------


category_id      0
category_name    0
dtype: int64


--------------------------Shape Of Data---------------------------------


(4, 2)

=================================Users Data=================================
-------------------------------TOP 5 RECORDS-----------------------------


user_id               email   username first_name last_name  \
0        1      john@gmail.com      johnd       john       doe   
1        2  morrison@gmail.com   mor_2314      david  morrison   
2        3     kevin@gmail.com  kevinryan      kevin      ryan   
3        4       don@gmail.com     donero        don     romer   
4        5     derek@gmail.com      derek      derek    powell   

        phone_num            street         city    zip_code     long      lat  
0  1-570-236-7033          new road     kilcoole  12926-3874  81.1496 -37.3159  
1  1-570-236-7033         Lovers Ln     kilcoole  12926-3874  81.1496 -37.3159  
2  1-567-094-1345        Frances Ct      Cullman  29567-1452 -30.1310  40.3467  
3  1-765-789-6734  Hunters Creek Dr  San Antonio  98234-1734 -20.1310  50.3467  
4  1-956-001-1945          adams St  san Antonio  80796-1234 -40.1310  40.3467


-------------------------------INFO--------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   user_id     10 non-null     int64  
 1   email       10 non-null     object 
 2   username    10 non-null     object 
 3   first_name  10 non-null     object 
 4   last_name   10 non-null     object 
 5   phone_num   10 non-null     object 
 6   street      10 non-null     object 
 7   city        10 non-null     object 
 8   zip_code    10 non-null     object 
 9   long        10 non-null     float64
 10  lat         10 non-null     float64
dtypes: float64(2), int64(1), object(8)
memory usage: 1012.0+ bytes


None


-------------------------------Describe----------------------------------


user_id       long        lat
count  10.00000  10.000000  10.000000
mean    5.50000   9.250758  20.763964
std     3.02765  43.122361  33.061214
min     1.00000 -40.131000 -37.315900
25%     3.25000 -20.441814  12.801125
50%     5.50000  -0.018500  35.186220
75%     7.75000  20.616900  40.346700
max    10.00000  81.149600  50.346700


-------------------------------Columns-----------------------------------


Index(['user_id', 'email', 'username', 'first_name', 'last_name', 'phone_num',
       'street', 'city', 'zip_code', 'long', 'lat'],
      dtype='object')


----------------------------Missing Values-------------------------------


user_id       0
email         0
username      0
first_name    0
last_name     0
phone_num     0
street        0
city          0
zip_code      0
long          0
lat           0
dtype: int64


--------------------------Shape Of Data---------------------------------


(10, 11)

=================================Sales Data=================================
-------------------------------TOP 5 RECORDS-----------------------------


sales_id  product_id  category_id  cart_id  user_id  quantity_purchased  \
0         1           1            3        1        1                   4   
1         2           2            3        1        1                   1   
2         3           3            3        1        1                   6   
3         4           2            3        2        1                   4   
4         5           1            3        2        1                  10   

   product_total_price  rating_count  rating_rate  total_cart_price  \
0               439.80           120          3.9            798.04   
1                22.30           259          4.1            798.04   
2               335.94           500          4.7            798.04   
3                89.20           259          4.1           2578.70   
4              1099.50           120          3.9           2578.70   

   distinct_products_in_cart  
0                          3  
1                          3  
2                          3  
3                          3  
4                          3


-------------------------------INFO--------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   sales_id                   14 non-null     int64  
 1   product_id                 14 non-null     int64  
 2   category_id                14 non-null     int64  
 3   cart_id                    14 non-null     int64  
 4   user_id                    14 non-null     int64  
 5   quantity_purchased         14 non-null     int64  
 6   product_total_price        14 non-null     float64
 7   rating_count               14 non-null     int64  
 8   rating_rate                14 non-null     float64
 9   total_cart_price           14 non-null     float64
 10  distinct_products_in_cart  14 non-null     int64  
dtypes: float64(3), int64(8)
memory usage: 1.3 KB


None


-------------------------------Describe----------------------------------


sales_id  product_id  category_id    cart_id    user_id  \
count   14.0000   14.000000    14.000000  14.000000  14.000000   
mean     7.5000    5.714286     2.428571   3.428571   2.500000   
std      4.1833    5.195095     0.937614   2.064882   1.951331   
min      1.0000    1.000000     1.000000   1.000000   1.000000   
25%      4.2500    1.250000     2.000000   2.000000   1.000000   
50%      7.5000    4.000000     3.000000   3.000000   2.000000   
75%     10.7500    8.750000     3.000000   5.000000   3.000000   
max     14.0000   18.000000     4.000000   7.000000   8.000000   

       quantity_purchased  product_total_price  rating_count  rating_rate  \
count           14.000000            14.000000     14.000000    14.000000   
mean             3.000000           335.090714    257.214286     3.835714   
std              2.541956           420.111207    147.945509     0.821450   
min              1.000000             9.850000    100.000000     1.900000   
25%              1.000000            32.725000    120.000000     3.450000   
50%              2.000000           218.950000    231.000000     3.900000   
75%              4.000000           415.350000    400.000000     4.475000   
max             10.000000          1390.000000    500.000000     4.800000   

       total_cart_price  distinct_products_in_cart  
count         14.000000                  14.000000  
mean         879.259286                   2.285714  
std          961.534347                   0.726273  
min            9.850000                   1.000000  
25%          283.900000                   2.000000  
50%          560.000000                   2.000000  
75%          798.040000                   3.000000  
max         2578.700000                   3.000000


-------------------------------Columns-----------------------------------


Index(['sales_id', 'product_id', 'category_id', 'cart_id', 'user_id',
       'quantity_purchased', 'product_total_price', 'rating_count',
       'rating_rate', 'total_cart_price', 'distinct_products_in_cart'],
      dtype='object')


----------------------------Missing Values-------------------------------


sales_id                     0
product_id                   0
category_id                  0
cart_id                      0
user_id                      0
quantity_purchased           0
product_total_price          0
rating_count                 0
rating_rate                  0
total_cart_price             0
distinct_products_in_cart    0
dtype: int64


--------------------------Shape Of Data---------------------------------


(14, 11)

**Key Observations for Products:**
- No missing values detected.
- The `price` column ranges from $7.95 to $999.99.
- The `rating` column needs further processing as it's stored as a dictionary.


### Data Cleaning & Transformation Overview
- Removed unnecessary columns like `__v` from the raw datasets.
- Flattened nested JSON structures such as `rating` in products and `address` in users.
- Standardized column names to match the database schema.


In [6]:
# Final check: Ensure no missing values and data types are correct
for df_name, df in [("Products", products), ("Categories", categories), ("Carts", carts), ("Users", users)]:
    print(f"\nFinal Check for {df_name}")
    print(df.isnull().sum())
    print(df.dtypes)


Final Check for Products
product_id      0
price           0
product_name    0
description     0
image_url       0
dtype: int64
product_id        int64
price           float64
product_name     object
description      object
image_url        object
dtype: object

Final Check for Categories
category_id      0
category_name    0
dtype: int64
category_id       int64
category_name    object
dtype: object

Final Check for Carts
cart_id      0
cart_date    0
dtype: int64
cart_id       int64
cart_date    object
dtype: object

Final Check for Users
user_id       0
email         0
username      0
first_name    0
last_name     0
phone_num     0
street        0
city          0
zip_code      0
long          0
lat           0
dtype: int64
user_id         int64
email          object
username       object
first_name     object
last_name      object
phone_num      object
street         object
city           object
zip_code       object
long          float64
lat           float64
dtype: object
